### PR TITLE
DigiByte

### PR DESCRIPTION
**Pull request by chappjc**
Resolves https://github.com/decred/dcrdex/issues/1992

Regnet cannot be mined too quickly.  https://twitter.com/chappjc/status/1601403517151113216?t=-H0D1YgQvpu2kvk5rMi8MQ&s=1

I still need to verify serialization is unchanged.

https://digibyte.org/docs/integrationguide.pdf

### CODE REVIEW

File: dex/testing/dgb/harness.sh [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+export GAMMA_ADDRESS="dgbrt1q46scevzm9hu06y0pnl8v2fxmll9v8k3e3lnd43"
+# Delta is a named wallet in the beta wallet directory.
+export DELTA_WALLET_SEED="ehA5EC6mbNvGCJ4HqtjNc822ojnU2bGRLoc3cZipoYFFYT7tYrq4"
+export DELTA_ADDRESS="dgbrt1qdgzj8guegzuyfupcvy3vjlpfxpv8rgruqtkndf"
+# Signal that the node needs to restart after encrypting wallet
+export RESTART_AFTER_ENCRYPT="1"
+# $1 is the node to create with. $2 is the wallet name
+export NEW_WALLET_CMD="./\$1 createwallet \$2"
+export BLOCKS_TO_MINE=120 # it chokes with high diff if you mine much more, but if you stop at 100, nothing much is mature it seems
+# Run the harness
```
Comment (buck54321) [inline dex/testing/dgb/harness.sh:26]: When I tested Digibyte long long ago, I had to use `export EXTRA_DAEMON_ARGS="--disabledandelion=1"` for the server to see mempool transactions. I don't know if that's still required.

Comment (chappjc) [reply dex/testing/dgb/harness.sh:26]: I did too.  Was planning to add that.  `gettxout` doesn't find things promptly, in cases, without that.

**Comment by JoeGruffins** [comment]
In the last commit, are any of the images besides `dgb.png` used?

**Comment by chappjc** [comment]
> In the last commit, are any of the images besides `dgb.png` used?

No I'll rm the ones we don't like.

> Should v7.17.2 be ok? Running on testnet not finding any blocks. Any tips?

https://github.com/DigiByte-Core/digibyte/releases/tag/v7.17.3

But there's a nasty bug that makes "Loading block index..." at startup take in excess of 24 hrs.

I have a patch branch here: https://github.com/DigiByte-Core/digibyte/pull/102
Mainnet does not seem to have this issue for the reasons described in the PR desc.

The testnet seeders have been down.  That's another thing I raised with their community.  For now, you'll have to replace the following lines:

https://github.com/DigiByte-Core/digibyte/blob/5ac15f556e2150c2a23f2b4e45f32fda4de3f9b6/src/chainparams.cpp#L414-L415

without the "seed." prefix to the host name.

Some other possible nodes to add manually are 91.197.230.186, 24.101.88.154, 47.75.38.245, 199.47.64.33, 8.214.25.169
Also not a problem with mainnet.

**Comment by JoeGruffins** [comment]
The harness seems to to be working on simnet. Not sure about testnet...

It looks like the dcrdex harness needs for dgb to be added and a market for it.

**Comment by chappjc** [comment]
Squashed in a few fixes: https://github.com/decred/dcrdex/compare/c828cc98add27773b0304c681ed2fc20ff27cf19..60d7ac475299ec38acbb35d59a6cd72ab7f39ffa
Server starts, but haven't tested trades yet.

**Comment by buck54321** [comment]
Things are looking good. We've added the remaining work to our next proposal and I anticipate we can finish things up within weeks. Getting a DigiByte market up and running is another hurdle. We're working on improving our market-making tools. The rudimentary tools being released in 0.6 (any day now) can help bootstrap a market, but provide few knobs for fine-tuned bot behavior. We'll keep at it and keep in touch.

**Comment by ycagel** [comment]
Thanks for the prompt reply @buck54321! I will work on getting the DigiByte community involved once DGB is live on the platform. I think there are quite a few people that are excited about Decred DEX. Cheers.

**Review by buck54321 — COMMENTED** [review]
I doubt it would work anyway, because of the segwit misalignment, but when trying to trade, I see `dgb wallet has no network peers (check your network or firewall)`.

File: client/webserver/site/src/img/coins/dgb2.png
[the remark this answers is not in the document]
Comment (chappjc) [reply client/webserver/site/src/img/coins/dgb2.png:1]: No, we just have to pick the ones we like.  Some other possibly relevant comments here: https://github.com/decred/dcrdex/pull/1993#issuecomment-1397012325
What network were you trying btw?

File: client/asset/dgb/dgb.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+		WalletInfo:          WalletInfo,
+		Symbol:              "dgb",
+		Logger:              logger,
+		Network:             network,
+		ChainParams:         params,
+		Ports:               ports,
+		DefaultFallbackFee:  defaultFee,
+		DefaultFeeRateLimit: defaultFeeRateLimit,
+		LegacyBalance:       true,
+		Segwit:              false,
```
[the remark this answers is not in the document]
Comment (chappjc) [reply client/asset/dgb/dgb.go:166]: Yah, absolutely segwit.  One of the first alts according to their integ guide.  Will get this PR ready soon.

File: client/webserver/site/src/img/coins/dgb2.png
[the remark this answers is not in the document]
Comment (buck54321) [reply client/webserver/site/src/img/coins/dgb2.png:1]: Simnet

[the remark this answers is not in the document]
Comment (chappjc) [reply client/webserver/site/src/img/coins/dgb2.png:1]: Gotcha.  Don't retest yet BTW.  I'm still updating the harness so the alpha/beta nodes connect to each other.

[the remark this answers is not in the document]
Comment (chappjc) [reply client/webserver/site/src/img/coins/dgb2.png:1]: k, simnet connectivity issue sorted.  Haven't tried a swap or anything else yet though:

image

**Comment by chappjc** [comment]
Need to disable the auto miner too

**Comment by chappjc** [comment]
image

image

**Comment by ycagel** [comment]
Amazing stuff @chappjc @buck54321! Are those screenshots actual orders? Was the other code fixes what you discussed before?

Cheers.

**Comment by chappjc** [comment]
> Amazing stuff @chappjc @buck54321! Are those screenshots actual orders?

We ran a few trades with this PR.  Looks good so far.

We put up the PR in December as a quick initial feasibility check, but today the Decred stakeholders approved a broader initiative that included Digibyte integration, so it got moving again.  https://proposals.decred.org/record/ca6b749

> Was the other code fixes what you discussed before?

We're still targeting the released versions of Digibyte (v7.17.4), which doesn't include the regtest/simnet changes to make mining easier, like the `easypow` switch, but that will be valuable going forward. We haven't evaluated those changes yet, but we will.

It would also be good if the [testnet issues](https://github.com/DigiByte-Core/digibyte/pull/102) were addressed, but we can work around it since it doesn't affect simnet or mainnet in practice.

I'll probably have some final questions before we release though.  Stuff related to fee rates, dust limits, tx/block serialization, dandelion, etc.  Will get back once more on the dex team have tested and reviewed.

**Comment by chappjc** [comment]
Fine on mainnet too

image

https://digiexplorer.info/tx/4da4eca50b54e67b3d2c5684938686b42385429fad6188b13e353bbad22aa877

**Review by buck54321 — CHANGES_REQUESTED** [review]

File: client/asset/dgb/dgb.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+		{
+			Key:         "txsplit",
+			DisplayName: "Pre-split funding inputs",
+			Description: "When placing an order, create a \"split\" transaction to fund the order without locking more of the wallet balance than " +
+				"necessary. Otherwise, excess funds may be reserved to fund the order until the first swap contract is broadcast " +
+				"during match settlement, or the order is canceled. This an extra transaction for which network mining fees are paid. " +
+				"Used only for standing-type orders, e.g. limit orders without immediate time-in-force.",
+			IsBoolean:    true,
+			DefaultValue: true, // low fee, fast chain
+		}, // no ExternalFeeEstimator, so no apifeefallback option
```
[the remark this answers is not in the document]
Comment (chappjc) [reply client/asset/dgb/dgb.go:56]: > `apiFeeFallback` is optional via argument to `CommonConfigOpts`.

I didn't feel like refactoring `CommonConfigOpts` with inputs for fee rate overrides so we don't use BTC's defaults.

> Don't we want redeemconftarget?

Yah, missed that.

[the remark this answers is not in the document]
Comment (chappjc) [reply client/asset/dgb/dgb.go:56]: Oh, I also enabled `txsplit` by default, unlike what `CommonConfigOpts` uses.  I suppose I can use `CommonConfigOpts` to makes something that I then modify in all these ways.  Cloning all this `Description` text isn't desirable after all.

[the remark this answers is not in the document]
Comment (buck54321) [reply client/asset/dgb/dgb.go:56]: > I suppose I can use `CommonConfigOpts` to makes something that I then modify in all these ways.

That sounds worse. Better the way you have it.

**Comment by chappjc** [comment]
Squashed the fixups.

**Comment by chappjc** [comment]
Revalidated trades again today.  Any lingering concerns?

**Comment by buck54321** [comment]
I'll review this tomorrow. I'm sure it's g2g.

**Review by buck54321 — APPROVED** [review]

File: client/asset/btc/btc.go [hunk 4]
```diff
@@ -4751,7 +4751,7 @@ func (btc *baseWallet) signTxAndAddChange(baseTx *wire.MsgTx, addr btcutil.Addre
 		// Add the change output.
 		vSize0 := btc.calcTxSize(baseTx)
 		baseTx.AddTxOut(changeOutput)
-		changeSize := dexbtc.MsgTxVBytes(baseTx) - vSize0   // may be dexbtc.P2WPKHOutputSize
+		changeSize := btc.calcTxSize(baseTx) - vSize0       // may be dexbtc.P2WPKHOutputSize
```
[the remark this answers is not in the document]
Comment (chappjc) [reply client/asset/btc/btc.go:4754]: Looks that way.  Low impact afaict.

**Comment by ycagel** [comment]
Thanks for all the hard work guys! How are things looking?

**Comment by buck54321** [comment]
Work is merged. Didn't make it into 0.6.0, but we already have an 0.6.1 working for a (hopefully) quick follow-up and DigiByte will be in that release.

**Comment by ycagel** [comment]
Looks like there is more progress here guys! Awesome job and thank you!

**Comment by ycagel** [comment]
I see 0.6.1 was just released. Great work. Is DigiByte confirmed in that release?

**Pull request merged**